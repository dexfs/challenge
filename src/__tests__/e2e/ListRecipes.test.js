require('dotenv-flow').config()
const axios = require('axios')
const MockAdapter = require('axios-mock-adapter')
const config = require('@config')()
const recipePuppyJson = require('../fixtures/recipepuppy.json')
const giphyJson = require('../fixtures/giphy.json')
const supertest = require('supertest')
const { app } = require('@bootstrap')

const startTestServer = async () => {
  return supertest(app)
}

const makeMockAxios = () => {
  const mockAxios = new MockAdapter(axios)
  return mockAxios
}

describe('ListRecipesAction', () => {
  let request
  beforeAll(async () => {
    request = await startTestServer()
  })
  it('should list recipes when ingredients passed', async (done) => {
    const query = 'onions,tomato'
    const mockAxios = makeMockAxios()
    const recipeRequest = `${config.apiRecipePuppyEndpoint}?i=${query}`
    mockAxios.onGet(recipeRequest).replyOnce(200, recipePuppyJson)
    mockAxios.onAny().reply(200, giphyJson)

    const response = await request.get(`/recipes/?i=${query}`)
    expect(response.body).toHaveProperty('keywords')
    expect(response.body).toHaveProperty('recipes')
    expect(response.body.keywords).toEqual(['onions', 'tomato'])
    expect(response.body).toMatchSnapshot()
    done()
  })
  it('should return 403 if Giphy API KEY no passed', async (done) => {
    const query = 'onions,tomato'
    const mockAxios = makeMockAxios()
    const recipeRequest = `${config.apiRecipePuppyEndpoint}?i=${query}`
    mockAxios.onGet(recipeRequest).replyOnce(200, {
      title: 'Recipe Puppy',
      version: 0.1,
      href: 'http://www.recipepuppy.com/',
      results: [
        {
          title: 'any_title',
          href: 'any_href',
          ingredients: 'any_ingredienes',
          thumbnail: 'any_thumbnail'
        }
      ]
    })

    mockAxios.onGet(config.apiGiphyEndpoint, {
      params: {
        api_key: config.apiGiphyKey,
        q: 'any_title'
      }
    }).replyOnce(403, 'Forbiden')
    const response = await request.get(`/recipes/?i=${query}`)
    expect(response.status).toBe(403)
    expect(response.body).toHaveProperty('status')
    expect(response.body).toHaveProperty('message')
    done()
  })

  it('should return 503 if Recipe Service is unavailable', async (done) => {
    const query = 'onions,tomato'
    const mockAxios = makeMockAxios()
    const recipeRequest = `${config.apiRecipePuppyEndpoint}?i=${query}`
    mockAxios.onGet(recipeRequest).networkError()
    const response = await request.get(`/recipes/?i=${query}`)
    console.log(response.body)
    expect(response.status).toBe(503)
    expect(response.body).toHaveProperty('status')
    expect(response.body).toHaveProperty('message')
    expect(response.body.message).toBe('Service Unavailable')
    done()
  })

  it('should return 403 if Giphy API KEY no passed', async (done) => {
    const query = 'onions,tomato'
    const mockAxios = makeMockAxios()
    const recipeRequest = `${config.apiRecipePuppyEndpoint}?i=${query}`
    mockAxios.onGet(recipeRequest).replyOnce(200, {
      title: 'Recipe Puppy',
      version: 0.1,
      href: 'http://www.recipepuppy.com/',
      results: [
        {
          title: 'any_title',
          href: 'any_href',
          ingredients: 'any_ingredienes',
          thumbnail: 'any_thumbnail'
        }
      ]
    })

    mockAxios.onGet(config.apiGiphyEndpoint, {
      params: {
        api_key: config.apiGiphyKey,
        q: 'any_title'
      }
    }).networkError()
    const response = await request.get(`/recipes/?i=${query}`)
    expect(response.status).toBe(503)
    expect(response.body).toHaveProperty('status')
    expect(response.body).toHaveProperty('message')
    expect(response.body.message).toBe('Service Unavailable')
    done()
  })
})
