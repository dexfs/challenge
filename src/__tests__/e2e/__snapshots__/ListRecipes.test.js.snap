// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListRecipesAction should list recipes when ingredients passed 1`] = `
Object {
  "keywords": Array [
    "onions",
    "tomato",
  ],
  "recipes": Array [
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "black pepper",
        "cream cheese",
        "eggs",
        "garlic",
        "italian seasoning",
        "milk",
        "olive oil",
        "onions",
        "parmesan cheese",
        "red pepper",
        "salt",
        "tomato",
        "vermicelli",
        "zucchini",
      ],
      "links": "http://find.myrecipes.com/recipes/recipefinder.dyn?action=displayRecipe&recipe_id=520763",
      "title": "Vegetable-Pasta Oven Omelet",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "bacon",
        "black pepper",
        "black pepper",
        "black pepper",
        "butter",
        "eggs",
        "garlic",
        "onions",
        "oregano",
        "roasted red peppers",
        "salt",
      ],
      "links": "http://www.bigoven.com/43919-Roasted-Pepper-and-Bacon-Omelet-recipe.html",
      "title": "Roasted Pepper and Bacon Omelet",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "basil",
        "broccoli",
        "eggs",
        "garlic",
        "lowfat milk",
        "onions",
        "parmesan cheese",
        "parmesan cheese",
        "salt",
        "tomato",
      ],
      "links": "http://cookeatshare.com/recipes/broccoli-oven-omelet-92851",
      "title": "
Broccoli Oven Omelet Recipe

",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "black pepper",
        "caraway seed",
        "coriander",
        "eggplant",
        "eggs",
        "garlic",
        "lemon",
        "olive oil",
        "onions",
        "salt",
      ],
      "links": "http://www.epicurious.com/recipes/food/views/Eggplant-Omelet-with-Coriander-and-Caraway-306",
      "title": "Eggplant Omelet with Coriander and Caraway",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "black pepper",
        "caraway seed",
        "coriander",
        "egg",
        "eggplant",
        "garlic",
        "lemon",
        "olive oil",
        "onions",
        "salt",
      ],
      "links": "http://www.bigoven.com/143831-Eggplant-Omelet-with-Coriander-And-Caraway-recipe.html",
      "title": "Eggplant Omelet with Coriander And Caraway",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "black pepper",
        "bread",
        "butter",
        "chicken broth",
        "cottage cheese",
        "egg substitute",
        "garlic",
        "mozzarella cheese",
        "onions",
        "oregano",
      ],
      "links": "http://www.bigoven.com/45151-Broccoli-and-Cheese-Omelet-recipe.html",
      "title": "Broccoli and Cheese Omelet",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "basil",
        "eggs",
        "garlic",
        "gruyere cheese",
        "leaves",
        "olive oil",
        "onions",
        "parmesan cheese",
        "parsley",
        "thyme",
      ],
      "links": "http://www.bigoven.com/149383-Chard-And-Onion-Omelet-recipe.html",
      "title": "Chard And Onion Omelet",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "eggs",
        "garlic",
        "olive oil",
        "onions",
        "parmesan cheese",
        "peas",
        "potato",
        "red pepper",
        "salt",
        "tomato",
        "zucchini",
      ],
      "links": "http://www.cdkitchen.com/recipes/recs/2184/Picnic-Omelet-Squares99498.shtml",
      "title": "Picnic Omelet Squares Recipe",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "black pepper",
        "chili powder",
        "eggs",
        "garlic",
        "green pepper",
        "milk",
        "mushroom",
        "onions",
        "salt",
        "vegetable oil",
        "water",
      ],
      "links": "http://www.cooks.com/rec/view/0,185,153160-249194,00.html",
      "title": "Spanish Omelet",
    },
    Object {
      "gif": "https://media2.giphy.com/media/gzE15pfW1jxkI/giphy.gif?cid=b4115905q0rx72iy2sfho383rnocdc8tftparnh7g29ktut1&rid=giphy.gif",
      "ingredients": Array [
        "almonds",
        "black pepper",
        "egg substitute",
        "eggs",
        "flour",
        "garlic",
        "goat cheese",
        "milk",
        "nonstick cooking spray",
        "olive oil",
        "onions",
        "parsley",
        "red wine vinegar",
        "salad greens",
        "salad greens",
        "salt",
        "thyme",
      ],
      "links": "http://find.myrecipes.com/recipes/recipefinder.dyn?action=displayRecipe&recipe_id=1622444",
      "title": "Onion and Fresh Herb Omelet with Mixed Greens",
    },
  ],
}
`;
